x-pack/plugin/index-lifecycle/src/main/java/org/elasticsearch/xpack/indexlifecycle/MaintenanceModeUpdateTask.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.indexlifecycle;
+
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.ClusterStateUpdateTask;
+import org.elasticsearch.cluster.metadata.MetaData;
+import org.elasticsearch.common.logging.ESLoggerFactory;
+import org.elasticsearch.common.util.set.Sets;
+import org.elasticsearch.xpack.core.indexlifecycle.IndexLifecycleMetadata;
+import org.elasticsearch.xpack.core.indexlifecycle.OperationMode;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+public class MaintenanceModeUpdateTask extends ClusterStateUpdateTask {
+    private static final Logger logger = ESLoggerFactory.getLogger(MaintenanceModeUpdateTask.class);
+    private final OperationMode mode;
+    private static final Map<OperationMode, Set<OperationMode>> VALID_MODE_CHANGES = new HashMap<>();
+
+    static {
+        VALID_MODE_CHANGES.put(OperationMode.NORMAL, Sets.newHashSet(OperationMode.MAINTENANCE_REQUESTED));
+        VALID_MODE_CHANGES.put(OperationMode.MAINTENANCE_REQUESTED, Sets.newHashSet(OperationMode.NORMAL, OperationMode.MAINTENANCE));
+        VALID_MODE_CHANGES.put(OperationMode.MAINTENANCE, Sets.newHashSet(OperationMode.NORMAL));
```

Comment:
Instead of having these here, we could utilise the OperationMode enum and pass the valid next modes into each enum via a constructor then we could have a `isValidChange(OperationMode)` method on it which check the mode in the parameter exists in the list of valid modes on that enum value. This will mean that the OperationMode enum completely owns the state transitions

Reply:
hmm, I tried this but it got just as complicated because the constructor elements cannot be forward references. and I preferred not to use strings.

Instead, I'll move all the logic to this `isValidChange` method